### ISSUE TITLE
Installation issue: netlib-lapack

### ISSUE THREAD

**Issue by quellyn** [LLNL/spack#34200]
### Steps to reproduce the issue

```console
$ spack spec -I netlib-lapack%xl_r ^cmake%gcc

Input spec
--------------------------------
 -   netlib-lapack%xl_r
 -       ^cmake%gcc

Concretized
--------------------------------
 -   netlib-lapack@3.10.1%xl_r@16.1 cflags="-F/projects/opt/ppc64le/ibm/xlc-16.1.1.11/xlC/16.1.1/etc/xlc.cfg.rhel.7.9.gcc.8.3.0.cuda.11.0 -Wl,-rpath,/projects/opt/ppc64le/ibm/xlc-16.1.1.11/xlC/16.1.1/lib -L/projects/opt/ppc64le/ibm/xlc-16.1.1.11/xlC/16.1.1/lib -Wl,-rpath,/projects/opt/ppc64le/ibm/xlc-16.1.1.11/lib -L/projects/opt/ppc64le/ibm/xlc-16.1.1.11/lib" cxxflags="-F/projects/opt/ppc64le/ibm/xlc-16.1.1.11/xlC/16.1.1/etc/xlc.cfg.rhel.7.9.gcc.8.3.0.cuda.11.0 -Wl,-rpath,/projects/opt/ppc64le/ibm/xlc-16.1.1.11/xlC/16.1.1/lib -L/projects/opt/ppc64le/ibm/xlc-16.1.1.11/xlC/16.1.1/lib -Wl,-rpath,/projects/opt/ppc64le/ibm/xlc-16.1.1.11/lib -L/projects/opt/ppc64le/ibm/xlc-16.1.1.11/lib" fflags="-F/projects/opt/ppc64le/ibm/xlf-16.1.1.11/xlf/16.1.1/etc/xlf.cfg.rhel.7.9.gcc.8.3.0.cuda.11.0 -Wl,-rpath,/projects/opt/ppc64le/ibm/xlf-16.1.1.11/xlf/16.1.1/lib -L/projects/opt/ppc64le/ibm/xlf-16.1.1.11/xlf/16.1.1/lib -Wl,-rpath,/projects/opt/ppc64le/ibm/xlf-16.1.1.11/lib -L/projects/opt/ppc64le/ibm/xlf-16.1.1.11/lib" ~external-blas~ipo+lapacke+shared~xblas build_system=cmake build_type=RelWithDebInfo arch=linux-rhel7-power9le
[+]      ^cmake@3.24.3%gcc@8.3.0 cflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" cxxflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" fflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" ~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=linux-rhel7-power9le
[+]          ^ncurses@6.3%gcc@8.3.0 cflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" cxxflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" fflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" ~symlinks+termlib abi=none build_system=autotools arch=linux-rhel7-power9le
[+]              ^gnuconfig@2021-08-14%gcc@8.3.0 cflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" cxxflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" fflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64"  build_system=generic arch=linux-rhel7-power9le
[+]              ^pkgconf@1.8.0%gcc@8.3.0 cflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" cxxflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" fflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64"  build_system=autotools arch=linux-rhel7-power9le
[+]          ^openssl@1.1.1s%gcc@8.3.0 cflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" cxxflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" fflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" ~docs~shared build_system=generic certs=mozilla arch=linux-rhel7-power9le
[+]              ^ca-certificates-mozilla@2022-10-11%gcc@8.3.0 cflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" cxxflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" fflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64"  build_system=generic arch=linux-rhel7-power9le
[+]              ^perl@5.36.0%gcc@8.3.0 cflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" cxxflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" fflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" +cpanm+shared+threads build_system=generic arch=linux-rhel7-power9le
[+]                  ^berkeley-db@18.1.40%gcc@8.3.0 cflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" cxxflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" fflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" +cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-rhel7-power9le
[+]                  ^bzip2@1.0.8%gcc@8.3.0 cflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" cxxflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" fflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" ~debug~pic+shared build_system=generic arch=linux-rhel7-power9le
[+]                      ^diffutils@3.8%gcc@8.3.0 cflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" cxxflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" fflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64"  build_system=autotools arch=linux-rhel7-power9le
[+]                          ^libiconv@1.16%gcc@8.3.0 cflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" cxxflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" fflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64"  build_system=autotools libs=shared,static arch=linux-rhel7-power9le
[+]                  ^gdbm@1.23%gcc@8.3.0 cflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" cxxflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" fflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64"  build_system=autotools arch=linux-rhel7-power9le
[+]                      ^readline@8.1.2%gcc@8.3.0 cflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" cxxflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" fflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64"  build_system=autotools arch=linux-rhel7-power9le
[+]              ^zlib@1.2.13%gcc@8.3.0 cflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" cxxflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" fflags="-Wl,-rpath,/projects/opt/ppc64le/p9/gcc/8.3.0/lib64" +optimize+pic+shared build_system=makefile arch=linux-rhel7-power9le
```

To reproduce:
1. Clone spack and checkout v0.19.0 release:
```
$ git clone git@github.com:spack/spack.git
$ cd spack
$ git checkout v0.19.0
```
2. Load a module for the IBM XL compiler:
```
$ module load ibm/xlc-16.1.1.11-xlf-16.1.1.11-gcc-8.3.0
```
3. Source spack setup script:
```
$ source ./share/spack/setup-env.sh
```
4. Detect compiler:
```
$ spack compiler find

==> Added 5 new compilers to /vast/home/quellyn/.spack/linux/compilers.yaml
    xl_r@16.1  xl@16.1  gcc@8.3.0  gcc@4.8.5  cce@16.01.0001.0011
==> Compilers are defined in the following files:
    /vast/home/quellyn/.spack/linux/compilers.yaml
```

NOTE the detection of a "`cce`" compiler above with the same release number as the `xl` package. Spack's compiler detection seems to have gone awry here:

```
$ spack compiler info cce

cce@16.01.0001.0011:
	paths:
		cc = /projects/opt/ppc64le/ibm/xlc-16.1.1.11/xlC/16.1.1/bin/cc
		cxx = None
		f77 = None
		fc = None
	modules  = []
	operating system  = rhel7
```

Other issues are that Spack isn't picking up the Fortran compiler paths for XL:
```
$ spack compiler info xl_r
xl_r@16.1:
	paths:
		cc = /projects/opt/ppc64le/ibm/xlc-16.1.1.11/xlC/16.1.1/bin/xlc_r
		cxx = /projects/opt/ppc64le/ibm/xlc-16.1.1.11/xlC/16.1.1/bin/xlC_r
		f77 = None
		fc = None
	flags:
		fflags = ['-qzerosize']
	modules  = []
	operating system  = rhel7
```

At this point, I ended up just dropping in my own `compilers.yaml` into `~/.spack/linux` (see attached).

5. Try to build netlib-lapack with XL:
```
$ spack install -v netlib-lapack%xl_r ^cmake%gcc
```

I believe this error stems from the recent adoption of #30738 , so I am mentioning @alalazo on this.

### Error message

<details><summary>Error message</summary><pre>
==> Installing netlib-lapack-3.10.1-n2jypgx6kzkoxiix35jity2vdy26spvi
==> No binary for netlib-lapack-3.10.1-n2jypgx6kzkoxiix35jity2vdy26spvi found: installing from source
==> Fetching https://mirror.spack.io/_source-cache/archive/cd/cd005cd021f144d7d5f7f33c943942db9f03a28d110d6a3b80d718a295f7f714.tar.gz
==> Ran patch() for netlib-lapack
==> netlib-lapack: Executing phase: 'cmake'
==> Error: AttributeError: 'CMakeBuilder' object has no attribute 'compiler'

The 'netlib-lapack' package cannot find an attribute while trying to build from sources. This might be due to a change in Spack's package format to support multiple build-systems for a single package. You can fix this by updating the build recipe, and you can also report the issue as a bug. More information at https://spack.readthedocs.io/en/latest/packaging_guide.html#installation-procedure

/vast/home/quellyn/GIT/spack/lib/spack/spack/build_environment.py:1086, in _setup_pkg_and_run:
       1083        tb_string = traceback.format_exc()
       1084
       1085        # build up some context from the offending package so we can
  >>   1086        # show that, too.
       1087        package_context = get_package_context(tb)
       1088
       1089        logfile = None

See build log for details:
  /tmp/quellyn/spack-stage/spack-stage-netlib-lapack-3.10.1-n2jypgx6kzkoxiix35jity2vdy26spvi/spack-build-out.txt
</pre></details>

### Information on your system

```
$ spack debug report
* **Spack:** 0.19.0
* **Python:** 3.6.8
* **Platform:** linux-rhel7-power9le
* **Concretizer:** clingo
```

### Additional information

[compilers.yaml.txt](https://github.com/spack/spack/files/10117761/compilers.yaml.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/10117762/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/10117763/spack-build-out.txt)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by quellyn** [comment]
I think this is definitely fallout from #30738; when I roll back the `netlib-lapack` `package.py` recipe to the commit before that, the package builds fine.

I have been able to build with GCC, so I'm guessing it's something XL specific.

**Comment by quellyn** [comment]
Has anyone had a chance to confirm this?

**Comment by quellyn** [comment]
Fixed by #35048.
